### PR TITLE
PCI VirtIO tranport layer testing and cleanup

### PR DESCRIPTION
**Pull request by bchalios**
## Changes

Add a few unit tests and handle some corner cases in the PCI VirtIO transport code 

## Reason

Be more spec compliant and increase testing coverage.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following Developer
Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [x] I have read and understand [CONTRIBUTING.md][3].
- [x] I have run `tools/devtool checkbuild --all` to verify that the PR passes
  build checks on all supported architectures.
- [x] I have run `tools/devtool checkstyle` to verify that the PR passes the
  automated style checks.
- [x] I have described what is done in these changes, why they are needed, and
  how they are solving the problem in a clear and encompassing way.
- [x] I have updated any relevant documentation (both in code and in the docs)
  in the PR.
- [x] I have mentioned all user-facing changes in `CHANGELOG.md`.
- [x] If a specific issue led to this PR, this PR closes the issue.
- [x] When making API changes, I have followed the
  [Runbook for Firecracker API changes][2].
- [x] I have tested all new and changed functionalities in unit tests and/or
  integration tests.
- [x] I have linked an issue to every new `TODO`.

______________________________________________________________________

- [x] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

### CODE REVIEW

File: src/vmm/src/devices/virtio/transport/pci/common_config.rs [hunk 1]
```diff
@@ -271,7 +199,11 @@ impl VirtioPciCommonConfig {
                 }
             }
             0x16 => self.queue_select = value,
-            0x18 => self.with_queue_mut(queues, |q| q.size = value),
+            0x18 => self.with_queue_mut(queues, |q| {
+                if value <= q.max_size {
+                    q.size = value
+                }
```
Comment (ShadowCurse) [inline src/vmm/src/devices/virtio/transport/pci/common_config.rs]: We have a check for the correct `size` of the queue in the `Queue::initialize`, so no need to do it here.

File: src/vmm/src/devices/virtio/transport/pci/device.rs [hunk 2]
```diff
@@ -107,8 +106,7 @@ impl VirtioPciCap {
                 + VIRTIO_PCI_CAP_LEN_OFFSET,
             cfg_type: cfg_type as u8,
             pci_bar: VIRTIO_BAR_INDEX,
-            id: 0,
-            padding: [0; 2],
```
Comment (ShadowCurse) [inline src/vmm/src/devices/virtio/transport/pci/device.rs]: Maybe I am missing smth but in the spec the definition is:
```c
struct virtio_pci_cap { 
        u8 cap_vndr;    /* Generic PCI field: PCI_CAP_ID_VNDR */ 
        u8 cap_next;    /* Generic PCI field: next ptr. */ 
        u8 cap_len;     /* Generic PCI field: capability length */ 
        u8 cfg_type;    /* Identifies the structure. */ 
        u8 bar;         /* Where to find it. */ 
        u8 id;          /* Multiple capabilities of the same type */ 
        u8 padding[2];  /* Pad to full dword. */ 
        le32 offset;    /* Offset within bar. */ 
        le32 length;    /* Length of the structure, in bytes. */ 
}; 
```
so `id` field does technically exist.

Comment (bchalios) [reply src/vmm/src/devices/virtio/transport/pci/device.rs]: Link? Can't see it here: https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-1090004

Comment (ShadowCurse) [reply src/vmm/src/devices/virtio/transport/pci/device.rs]: I am looking at 1.3 version: https://docs.oasis-open.org/virtio/virtio/v1.3/csd01/virtio-v1.3-csd01.html#x1-1300001

File: src/vmm/src/devices/virtio/transport/pci/common_config.rs (hunk 1 above)
Comment (bchalios) [reply src/vmm/src/devices/virtio/transport/pci/common_config.rs]: true. Hadn't thought of that. I guess it's a question of causing the device to fail if such a value comes (current behaviour) or just ignore it and default to `max_size`. WDYT?

File: src/vmm/src/devices/virtio/transport/pci/device.rs (hunk 2 above)
Comment (bchalios) [reply src/vmm/src/devices/virtio/transport/pci/device.rs]: oh cool! I'll revert to that.

Comment (bchalios) [reply src/vmm/src/devices/virtio/transport/pci/device.rs]: done!

File: src/vmm/src/devices/virtio/transport/pci/common_config.rs (hunk 1 above)
Comment (ShadowCurse) [reply src/vmm/src/devices/virtio/transport/pci/common_config.rs]: I think the current way we do it is fine. It is much easier to keep all logic that checks for incorrect device configuration in one place.

Comment (Manciukic) [reply src/vmm/src/devices/virtio/transport/pci/common_config.rs]: According to the spec, the device should read the max first, then optionally write a smaller value. Looking at the Linux driver, it never checks back that the size it configured was accepted by the device, and just continues with its life assuming the value it wrote was accepted. So I think in this case we can fail the device as the driver is not following the spec.

Comment (bchalios) [reply src/vmm/src/devices/virtio/transport/pci/common_config.rs]: fine by me. dropping this from here.

Comment (bchalios) [reply src/vmm/src/devices/virtio/transport/pci/common_config.rs]: done

**Review by ShadowCurse — APPROVED** [review]
